### PR TITLE
feat: unified sentiment processing pipeline with confidence scoring and usability mapping

### LINKED ISSUE #1988: Feature: Unified Sentiment Processing Pipeline with Confidence Estimation and Usability-Oriented Mapping -GSOC-2026
<html>
<body>
<html><head></head><body><h1>Unified Sentiment Processing Pipeline with Confidence Estimation and Usability-Oriented Mapping - (Phase-1)</h1>
<p><strong>Labels:</strong> <code>enhancement</code> <code>gsoc-2026</code> <code>sentiment analysis</code>
<hr>
<h2>What needs to be built</h2>

<h3>1. Standard output schema</h3>
<p>
All modules (text, audio, facial) should return results in a common format. Each output must include:
</p>
<ul>
  <li><code>emotion</code> - predicted label</li>
  <li><code>confidence</code> - score between 0 and 1</li>
  <li><code>usability_label</code> - mapped UX-focused category (e.g. frustration, satisfaction)</li>
  <li><code>modality</code> - source (<code>text</code>, <code>audio</code>, <code>face</code>, <code>fused</code>)</li>
  <li><code>raw_scores</code> - optional probabilities for all classes</li>
</ul>

<h3>2. Confidence extraction</h3>
<p>
Model outputs (logits) should be converted into probabilities using <strong>softmax</strong>.
The confidence score must always be the probability of the top prediction and not raw values.
</p>
<p>
This will be important later when combining multiple modalities.
</p>

<h3>3. Emotion to usability mapping</h3>
<p>
We need a simple mapping from generic emotion labels to usability-focused categories.
</p>
<ul>
  <li>Clear and readable (dictionary-based)</li>
  <li>Easy to extend</li>
  <li>Separate from core logic</li>
</ul>

<table>
  <thead>
    <tr>
      <th>Emotion</th>
      <th>Usability</th>
    </tr>
  </thead>
  <tbody>
    <tr><td>anger, disgust, sadness</td><td>frustration</td></tr>
    <tr><td>fear</td><td>cognitive_load</td></tr>
    <tr><td>joy</td><td>satisfaction</td></tr>
    <tr><td>surprise</td><td>engagement</td></tr>
    <tr><td>neutral</td><td>neutral</td></tr>
    <tr><td>confusion</td><td>confusion</td></tr>
  </tbody>
</table>

<hr>
<h2>Expected output format</h2>
<pre><code class="language-json">{
  "emotion": "confusion",
  "confidence": 0.81,
  "usability_label": "cognitive_load",
  "modality": "text",
  "raw_scores": {
    "confusion": 0.81,
    "neutral": 0.12,
    "anger": 0.04,
    "joy": 0.03
  }
}
</code></pre>
<hr></html>
</body>
</html>

### PR DESCRIPTION
**Pull request by HITESH-S-P**
**Closes #1988**
**Branch:** `feature/unified-sentiment-pipeline`
**Labels:** `gsoc-2026`
**Project:** `Multimodal Sentiment Analysis Engine for Usability Studies`

---

## Outcomes of this PR

This adds the foundational layer that every other module in the sentiment engine will build on. Before this, there was no standard contract for what a sentiment prediction should look like coming out of any module. This PR Fixes #1988 .

Three things went in:

**Output schema** — `schemas/output_schema.py` defines a Pydantic model (`SentimentOutput`) that all modules are expected to return. The fields are `emotion`, `confidence`, `usability_label`, `modality`, `raw_scores`, and `contributions` (the last two are optional and used downstream). Having this as a Pydantic model means bad outputs fail loudly at validation time rather than silently producing garbage in the fusion step.

**Emotion mapper** — `utils/emotion_mapper.py` handles the translation from generic model labels (anger, joy, fear, etc.) to usability-oriented categories (frustration, satisfaction, cognitive_load, etc.). It's just a dictionary lookup with a sensible fallback to `neutral`. The mapping is based on the RUXAILAB usability taxonomy. It's deliberately simple — a function, not a class — because there's no state to manage here and it should be easy for researchers to edit without reading much code.

**Sentiment processor** — `pipeline/sentiment_processor.py` is the wrapper that takes raw model output (a dict of label → logit/score), runs softmax normalization to get calibrated probabilities, picks the top prediction, attaches confidence, calls the mapper, and returns a validated `SentimentOutput`. Every module (text, audio, face) will instantiate one of these and call `.process()`.

---

## How to test this locally

```bash
cd sentiment-engine
pip install -r requirements.txt
pytest tests/test_sentiment_processor.py -v
```

All 9 tests should pass. There's one intentional edge case test for when all logits are equal — confidence should land at `1/n` for each class and the mapper should fall back gracefully.

---

## Sample output

```json
{
  "emotion": "anger",
  "confidence": 0.7023,
  "usability_label": "frustration",
  "modality": "text",
  "raw_scores": {
    "anger": 0.7023,
    "confusion": 0.1941,
    "neutral": 0.0742,
    "joy": 0.0294
  },
  "contributions": null
}
```

---

## Tests output

Screenshot 2026-03-29 205942

---

## Files changed

```
schemas/output_schema.py      
utils/emotion_mapper.py        
pipeline/sentiment_processor.py  
tests/test_sentiment_processor.py 
requirements.txt                 
```

---

This PR is part of my GSoC 2026 application for the Multimodal Sentiment Analysis Engine project.

### CODE REVIEW

**Comment by HITESH-S-P** [comment]
Hi, just following up on this PR. Would appreciate a review when you have time.
